### PR TITLE
Improve combinations

### PR DESCRIPTION
**Pull request by 8d1h**
The `combinations` function was very slow...
```julia
julia> @btime AbstractAlgebra.combinations(10, 3); # before
  32.587 μs (996 allocations: 92.59 KiB)

julia> @btime AbstractAlgebra.combinations(10, 3); # after
  5.971 μs (129 allocations: 15.47 KiB)

julia> @btime collect(Combinatorics.combinations(1:10, 3));
  5.783 μs (242 allocations: 18.05 KiB)
```
The output order of the combinations is same as before.

### CODE REVIEW

File: src/Matrix.jl [hunk 1]
```diff
@@ -1918,21 +1918,24 @@ end
...
     combinations(n::Int, k::Int)
+    combinations(v::AbstractVector, k::Int)
 
-Return an array consisting of k-combinations of {1,...,n} as arrays.
+Return an array consisting of k-combinations of {1,...,n} (or a given vector v)
+as arrays.
 """
-function combinations(n::Int, k::Int)
-   ans = Vector{Int}[]
+function combinations(n::Int, k::Int) combinations(1:n, k) end
```
Comment (thofma) [inline src/Matrix.jl:1926]:
```suggestion
combinations(n::Int, k::Int) = combinations(1:n, k)

```

File: src/Matrix.jl [hunk 2]
```diff
@@ -1923,7 +1923,7 @@ end
 Return an array consisting of k-combinations of {1,...,n} (or a given vector v)
 as arrays.
 """
-function combinations(n::Int, k::Int) combinations(1:n, k) end
+combinations(n::Int, k::Int) = combinations(1:n, k)
```
Comment (thofma) [inline src/Matrix.jl:1926]: This is adding the docstring to the one method.

One idea is to do
```julia
@doc Markdown.doc"""
    combinations(n::Int, k::Int)
    combinations(v::AbstractVector, k::Int)

Return an array consisting of k-combinations of {1,...,n} (or a given vector v)
as arrays.
"""
function combinations end
```

```

Comment (8d1h) [reply src/Matrix.jl:1926]: Right, thanks! I think I'll just add docstrings for both methods.

File: src/Matrix.jl [hunk 3]
```diff
@@ -1937,7 +1937,7 @@ function combinations(v::AbstractVector{T}, k::Int) where T
 end
 function _combinations_dfs!(ans::Vector{Vector{T}}, comb::Vector{T}, v::AbstractVector{T}, n::Int, k::Int) where T
    k < 1 && (pushfirst!(ans, comb[:]); return)
-   for m in n:-1:1
+   for m in n:-1:k
```
Comment (8d1h) [inline src/Matrix.jl:1940]: Sorry I was being stupid...

Comment (thofma) [reply src/Matrix.jl:1940]: Should there have been a test catching this?

Comment (8d1h) [reply src/Matrix.jl:1940]: No the result is correct, it's just a lot of redundant search (happens for example for `combinations(100,99)`, which is how I caught it).

**Pull request merged**